File: matlab/kernel_invocations_syrk.m
function kernel_invocations_syrk(A, C, b)

reps = str2num(getenv('LAMP_REPS'));

implicit_syrk_ = @() implicit_syrk(A, C);
syrk_implicit_noup_ = @() syrk_implicit_noup(A);

b.benchmark("syrk_implicit", reps, implicit_syrk_);
b.benchmark("syrk_implicit_noup", reps, syrk_implicit_noup_);
end

function [C, time] = implicit_syrk(A, C)
tic;
C = A*A' + C;
time = toc;
end

function [C, time] = syrk_implicit_noup(A)
tic;
C = A*A';
time = toc;
end

File: r/kernel_invocations_syrk.R
library("sets")

syrk_implicit <- function(A, C){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    ct <- matrix(C, ncol=ncol(C), nrow=nrow(C))
    cache_scrub()
    start.time <- Sys.time()
    ct <- at %*% t(at) + ct
    end.time <- Sys.time()
    res = tuple(as.double(end.time-start.time), ct)

    return(res)
}

syrk_implicit_noup <- function(A, C){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    ct <- matrix(C, ncol=ncol(C), nrow=nrow(C))
    cache_scrub()
    start.time <- Sys.time()
    ct <- at %*% t(at)
    end.time <- Sys.time()
    res = tuple(as.double(end.time-start.time), ct)

    return(res)
}

kernel_invocations_syrk <- function(b, ...){
    res1 <- benchmark('syrk_implicit', syrk_implicit, ...)
    res2 <- benchmark('syrk_implicit_noup', syrk_implicit_noup, ...)
}
